Data/ByteString.hs

```diff
@@ -1642,18 +1642,22 @@ breakSubstring pat =
 -- excess elements of the longer ByteString are discarded. This is
 -- equivalent to a pair of 'unpack' operations.
 zip :: ByteString -> ByteString -> [(Word8,Word8)]
-zip ps qs
-    | null ps || null qs = []
-    | otherwise = (unsafeHead ps, unsafeHead qs) : zip (unsafeTail ps) (unsafeTail qs)
+zip ps qs = case uncons ps of
+  Nothing         -> []
+  Just (psH, psT) -> case uncons qs of
+    Nothing         -> []
+    Just (qsH, qsT) -> (psH, qsH) : zip psT qsT
```

Comment:
Could you possibly dump Core and check that `zip` remained the same after this change?
```
{-# OPTIONS_GHC -ddump-simpl -dsuppress-all -dno-suppress-type-signatures #-}
```

Reply:
That was cool never done that before!
old:
```
-- RHS size: {terms: 10, types: 6, coercions: 0, joins: 0/0}
zip :: ByteString -> ByteString -> [(Word8, Word8)]
zip
  = \ (w_slSN :: ByteString) (w1_slSO :: ByteString) ->
      case w_slSN of { BS ww1_slSR ww2_slSS ww3_slST ->
      $wzip ww1_slSR ww2_slSS ww3_slST w1_slSO
      }
```
new:
```
-- RHS size: {terms: 10, types: 6, coercions: 0, joins: 0/0}
zip :: ByteString -> ByteString -> [(Word8, Word8)]
zip
  = \ (w_sm0D :: ByteString) (w1_sm0E :: ByteString) ->
      case w_sm0D of { BS ww1_sm0H ww2_sm0I ww3_sm0J ->
      $wzip ww1_sm0H ww2_sm0I ww3_sm0J w1_sm0E
      }
```

The naming seems different, not sure if that make an impact?

Reply:
Naming is irrelevant. 
So far so good, but could you please check definitions of `$wzip` in both cases as well?

Reply:
they look the same too, guess it wouldn't, make any difference in terms of performance then?
Old:
```
-- RHS size: {terms: 64, types: 53, coercions: 0, joins: 0/0}
$wzip
  :: Addr#
     -> ForeignPtrContents -> Int# -> ByteString -> [(Word8, Word8)]
$wzip
  = \ (ww_slSR :: Addr#)
      (ww1_slSS :: ForeignPtrContents)
      (ww2_slST :: Int#)
      (w_slSO :: ByteString) ->
      case <=# ww2_slST 0# of {
        __DEFAULT ->
          case w_slSO of { BS dt_XbZl dt1_XbZn dt2_XbZp ->
          case <=# dt2_XbZp 0# of {
            __DEFAULT ->
              : (case readWord8OffAddr# ww_slSR 0# realWorld# of
                 { (# ipv_ibCO, ipv1_ibCP #) ->
                 case touch# ww1_slSS ipv_ibCO of { __DEFAULT -> W8# ipv1_ibCP }
                 },
                 case readWord8OffAddr# dt_XbZl 0# realWorld# of
                 { (# ipv_ibCO, ipv1_ibCP #) ->
                 case touch# dt1_XbZn ipv_ibCO of { __DEFAULT -> W8# ipv1_ibCP }
                 })
                (zip_$s$wzip
                   (plusAddr# dt_XbZl 1#)
                   dt1_XbZn
                   (-# dt2_XbZp 1#)
                   (-# ww2_slST 1#)
                   ww1_slSS
                   (plusAddr# ww_slSR 1#));
            1# -> []
          }
          };
        1# -> []
      }
```
New:
```
-- RHS size: {terms: 64, types: 53, coercions: 0, joins: 0/0}
$wzip
  :: Addr#
     -> ForeignPtrContents -> Int# -> ByteString -> [(Word8, Word8)]
$wzip
  = \ (ww_sm0H :: Addr#)
      (ww1_sm0I :: ForeignPtrContents)
      (ww2_sm0J :: Int#)
      (w_sm0E :: ByteString) ->
      case <=# ww2_sm0J 0# of {
        __DEFAULT ->
          case w_sm0E of { BS dt_Xc1T dt1_Xc1V dt2_Xc1X ->
          case <=# dt2_Xc1X 0# of {
            __DEFAULT ->
              : (case readWord8OffAddr# ww_sm0H 0# realWorld# of
                 { (# ipv_ibYY, ipv1_ibYZ #) ->
                 case touch# ww1_sm0I ipv_ibYY of { __DEFAULT -> W8# ipv1_ibYZ }
                 },
                 case readWord8OffAddr# dt_Xc1T 0# realWorld# of
                 { (# ipv_ibYY, ipv1_ibYZ #) ->
                 case touch# dt1_Xc1V ipv_ibYY of { __DEFAULT -> W8# ipv1_ibYZ }
                 })
                (zip_$s$wzip
                   (plusAddr# dt_Xc1T 1#)
                   dt1_Xc1V
                   (-# dt2_Xc1X 1#)
                   (-# ww2_sm0J 1#)
                   ww1_sm0I
                   (plusAddr# ww_sm0H 1#));
            1# -> []
          }
          };
        1# -> []
      }
```